gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobufdataset.cpp

```diff
@@ -0,0 +1,302 @@
+            CPLDebug("FlatGeobuf", "Verified magicbytes");
+            return TRUE;
+        } else {
+            CPLError(CE_Failure, CPLE_OpenFailed,
+                "Unsupported FlatGeobuf version %d.\n",
+                poOpenInfo->pabyHeader[3]);
+        }
+
+    }
+
+    return FALSE;
+}
+
+void RegisterOGRFlatGeobuf()
+{
+    if( GDALGetDriverByName("FlatGeobuf") != NULL )
+        return;
+
+    GDALDriver *poDriver = new GDALDriver();
+    poDriver->SetDescription("FlatGeobuf");
+    poDriver->SetMetadataItem(GDAL_DCAP_VECTOR, "YES");
+    poDriver->SetMetadataItem(GDAL_DMD_LONGNAME, "FlatGeobuf");
+    poDriver->SetMetadataItem(GDAL_DMD_EXTENSION, "fgb");
+    poDriver->SetMetadataItem(GDAL_DMD_HELPTOPIC, "drv_flatgeobuf.html");
+    poDriver->SetMetadataItem(GDAL_DCAP_VIRTUALIO, "YES");
+    poDriver->SetMetadataItem( GDAL_DMD_CREATIONFIELDDATATYPES,
+                               "Integer Integer64 Real String IntegerList "
+                               "Integer64List RealList StringList" );
+    poDriver->SetMetadataItem( GDAL_DMD_CREATIONFIELDDATASUBTYPES, "Boolean" );
+
+    poDriver->pfnOpen = OGRFlatGeobufDataset::Open;
+    poDriver->pfnCreate = OGRFlatGeobufDataset::Create;
+    poDriver->pfnIdentify = OGRFlatGeobufDriverIdentify;
+
+    GetGDALDriverManager()->RegisterDriver(poDriver);
+}
+
+
+/************************************************************************/
+/*                          OGRFlatGeobufDataset()                          */
+/************************************************************************/
+
+OGRFlatGeobufDataset::OGRFlatGeobufDataset()
+{
+
+}
+
+OGRFlatGeobufDataset::OGRFlatGeobufDataset(const char *osDirName)
+{
+    CPLDebug("FlatGeobuf", "Request to create dataset %s", osDirName);
+    m_create = true;
+    m_pszName = CPLStrdup(osDirName);
+}
+
+/************************************************************************/
+/*                         ~OGRFlatGeobufDataset()                          */
+/************************************************************************/
+
+OGRFlatGeobufDataset::~OGRFlatGeobufDataset()
+{
+}
+
+/************************************************************************/
+/*                                Open()                                */
+/************************************************************************/
+
+GDALDataset *OGRFlatGeobufDataset::Open(GDALOpenInfo* poOpenInfo)
+{
+    if( !OGRFlatGeobufDriverIdentify(poOpenInfo) || poOpenInfo->eAccess == GA_Update )
+        return nullptr;
+
+    VSILFILE *fp = poOpenInfo->fpL;
+    CPLString osFilename(poOpenInfo->pszFilename);
+
+    uint64_t offset = sizeof(magicbytes);
+    CPLDebug("FlatGeobuf", "Start at offset (%lu)", sizeof(magicbytes));
+    VSIFSeekL(fp, offset, SEEK_SET);
+    uint32_t headerSize;
+    VSIFReadL(&headerSize, 4, 1, fp);
+    CPLDebug("FlatGeobuf", "headerSize (%d)", headerSize);
+    GByte* buf = static_cast<GByte*>(VSI_MALLOC_VERBOSE(headerSize));
```

Comment:
some validation of headerSize value would be welcome so that corrupted/hostile files don't blow up memory
buf != nullptr should be checked